src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphOperations.cs

```diff
@@ -176,179 +206,460 @@ private static bool TryResolveConflicts<TItem>(this GraphNode<TItem> root, List<
+        {
+            var count = innerNodes.Count;
+            for (var i = 0; i < count; i++)
+            {
+                var innerNode = innerNodes[i];
+                queue.Enqueue(NodeWithState.Create(innerNode, innerState));
+            }
+        }
+
+        private static void AddInnerNodesToQueue<TItem>(IList<GraphNode<TItem>> innerNodes, Queue<GraphNode<TItem>> queue)
+        {
+            var count = innerNodes.Count;
+            for (var i = 0; i < count; i++)
+            {
+                var innerNode = innerNodes[i];
+                queue.Enqueue(innerNode);
+            }
+        }
+
+        [ThreadStatic]
+        private static Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> _tempDowngrades;
+
+        public static Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> RentDowngradesDictionary()
+        {
+            var dictionary = _tempDowngrades;
+            if (dictionary != null)
+            {
+                _tempDowngrades = null;
+                return dictionary;
+            }
+
+            return new Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>>();
+        }
+
+        public static void ReleaseDowngradesDictionary(Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> dictionary)
+        {
+            if (_tempDowngrades == null)
+            {
+                dictionary.Clear();
+                _tempDowngrades = dictionary;
+            }
+        }
+
+        private static class Cache<TItem, TState>
+        {
+            [ThreadStatic]
+            private static Queue<NodeWithState<TItem, TState>> _queue;
+
+
+            public static Queue<NodeWithState<TItem, TState>> RentQueue()
+            {
+                var queue = _queue;
+                if (queue != null)
                 {
-                    visitor(node);
-                    return 0;
-                });
+                    _queue = null;
+                    return queue;
+                }
+
+                return new Queue<NodeWithState<TItem, TState>>();
+            }
+
+            public static void ReleaseQueue(Queue<NodeWithState<TItem, TState>> queue)
+            {
+                if (_queue == null)
+                {
+                    queue.Clear();
+                    _queue = queue;
+                }
+            }
+
+        }
+
+        private static class Cache<TItem>
+        {
+            [ThreadStatic]
+            private static Queue<GraphNode<TItem>> _queue;
+            [ThreadStatic]
+            private static Dictionary<string, GraphNode<TItem>> _dictionary;
+            [ThreadStatic]
+            private static Tracker<TItem> _tracker;
+
+            public static Queue<GraphNode<TItem>> RentQueue()
+            {
+                var queue = _queue;
+                if (queue != null)
+                {
+                    _queue = null;
+                    return queue;
```

Comment:
Yep, ThreadStatic is per thread. So it acquires it and sets it to null while only that thread has access to it. 

If it switched thread while it had it (e.g. via an await); then it a different action on that thread would see it as null and create a new one. Though there is nothing that would cause thread switching - just doing it to be safe.